webrtc/RTCDataChannel-bufferedAmount.html

```diff
@@ -5,165 +5,239 @@
-      assert_equals(channel1.bufferedAmount, helloString.length,
-        'Expect bufferedAmount to be the total length of all messages queued to send');
-
-      channel1.send(unicodeString);
-      assert_equals(channel1.bufferedAmount,
-        helloString.length + unicodeBuffer.byteLength,
-        'Expect bufferedAmount to be the total length of all messages queued to send');
-
-      channel1.send(helloBlob);
-      assert_equals(channel1.bufferedAmount,
-        helloString.length*2 + unicodeBuffer.byteLength,
-        'Expect bufferedAmount to be the total length of all messages queued to send');
-
-    }).catch(t.step_func(err =>
-      assert_unreached(`Unexpected promise rejection: ${err}`)));
-  }, 'bufferedAmount should increase by byte length for each message sent');
+
+    bufferedAmount
+      The bufferedAmount attribute must return the number of bytes of application
+      data (UTF-8 text and binary data) that have been queued using send() but that,
+      as of the last time the event loop started executing a task, had not yet been
+      transmitted to the network. (This thus includes any text sent during the
+      execution of the current task, regardless of whether the user agent is able
+      to transmit text asynchronously with script execution.) This does not include
+      framing overhead incurred by the protocol, or buffering done by the operating
+      system or network hardware. The value of the [[BufferedAmount]] slot will only
+      increase with each call to the send() method as long as the [[ReadyState]] slot
+      is open; however, the slot does not reset to zero once the channel closes. When
+      the underlying data transport sends data from its queue, the user agent MUST
+      queue a task that reduces [[BufferedAmount]] with the number of bytes that was
+      sent.
+
+
+  [WebMessaging]
+  interface MessageEvent : Event {
+    readonly attribute any data;
+    ...
+  };
+ */
+
+// Simple ASCII encoded string
+const helloString = 'hello';
+// ASCII encoded buffer representation of the string
+const helloBuffer = Uint8Array.of(0x68, 0x65, 0x6c, 0x6c, 0x6f);
+const helloBlob = new Blob([helloBuffer]);
+
+// Unicode string with multiple code units
+const unicodeString = '世界你好';
+// UTF-8 encoded buffer representation of the string
+const unicodeBuffer = Uint8Array.of(
+  0xe4, 0xb8, 0x96, 0xe7, 0x95, 0x8c,
+  0xe4, 0xbd, 0xa0, 0xe5, 0xa5, 0xbd);
+
+/*
+  Ensure .bufferedAmount is 0 initially for both sides.
+ */
+promise_test(async (t) => {
+  const pc1 = new RTCPeerConnection();
+  const pc2 = new RTCPeerConnection();
+  t.add_cleanup(() => pc1.close());
+  t.add_cleanup(() => pc2.close());
+
+  const [dc1, dc2] = await createDataChannelPair(pc1, pc2);
+
+  assert_equals(dc1.bufferedAmount, 0, 'Expect bufferedAmount to be 0');
+  assert_equals(dc2.bufferedAmount, 0, 'Expect bufferedAmount to be 0');
+}, 'bufferedAmount initial value should be 0 for both peers', {
+  timeout: 5000
```

Comment:
Even for async tests, I am not sure this is worth the effort to add specific timeouts.
In some configurations, 5 seconds will be what is expected to run the tests of the whole file for instance.

If we really want to do something there, maybe createDataChannelPair should actually reject after some timeout?
That way, if we have several await in a promise test, we will know which one is timing out.

Reply:
If we don't specify the local timeout, the test case will block running all other test cases. The alternative is to carefully wrap all possible awaitables with timeouts but I don't want to go that way for several reasons:

- it's quite a lot of boilerplate,
- makes the test cases brittle since people (including me) will tend to forget an awaitable here and there, and
- not every awaitable that could in theory time out is likely to time out.

Adding failure context to some awaitables by using timeouts is a good idea, though (and I have done that in vital places). Rejecting `createDataChannelPair` after some timeout could be done but since that isn't part of the test case definition, the question is when the timeout should fire...

Reply:
What about the tests that share 99% of their code?

Reply:
Sure, if some tests share 99%, it is ok to group them (and incur the timeout risk).

The alternative is of course to have some helper functions in some JS files.
That way, you keep the meat of each subtest in each promise_test function.
And you can then split tests as desired.

In general, the more tests on a given file, the more flaky.
And the more difficult to debug as well...